src/app/MessageDef.h

```diff
@@ -0,0 +1,1844 @@
+/**
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2016-2017 Nest Labs, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+/**
+ *    @file
+ *      This file defines CHIP interaction model message parsers and encoders
+ *      Interaction model profile.
+ *
+ */
+
+#ifndef _CHIP_INTERACTION_MODEL_MESSAGE_DEF_H
+#define _CHIP_INTERACTION_MODEL_MESSAGE_DEF_H
+
+#include <core/CHIPCore.h>
+#include <core/CHIPTLV.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+#include <util/basic-types.h>
+
+namespace chip {
+namespace app {
+/**
+ *  @brief
+ *    The CHIP interaction model message types.
+ *
+ *  These values are called out in CHIP Interaction Model: Encoding Specification
+ *
+ */
+enum
+{
+    kMsgType_SubscribeRequest      = 0x01,
+    kMsgType_ReadRequest           = 0x02,
+    kMsgType_ReportData            = 0x03,
+    kMsgType_WriteRequest          = 0x04,
+    kMsgType_WriteResponse         = 0x05,
+    kMsgType_InvokeCommandRequest  = 0x06,
+    kMsgType_InvokeCommandResponse = 0x07,
+};
+
+class ParserBase
+{
+public:
+    /**
+     *  @brief Initialize a TLVReader to point to the beginning of any tagged element in this request
+     *
+     *  @param [in]  aTagToFind Tag to find in the request
+     *  @param [out] apReader   A pointer to TLVReader, which will be initialized at the specified TLV element
+     *                          on success
+     *
+     *  @return #CHIP_NO_ERROR on success
+     */
+    CHIP_ERROR GetReaderOnTag(const uint64_t aTagToFind, chip::TLV::TLVReader * const apReader) const;
+
+protected:
+    chip::TLV::TLVReader mReader;
+
+    ParserBase();
+
+    template <typename T>
+    CHIP_ERROR GetUnsignedInteger(const uint8_t aContextTag, T * const apLValue) const;
+
+    template <typename T>
+    CHIP_ERROR GetSimpleValue(const uint8_t aContextTag, const chip::TLV::TLVType aTLVType, T * const apLValue) const;
+};
+
+class ListParserBase : public ParserBase
+{
+protected:
+    ListParserBase();
+
+public:
+    /**
+     *  @brief Initialize the parser object with TLVReader
+     *
+     *  @param [in] aReader A pointer to a TLVReader, which should be on the element of the array element
+     *
+     *  @return #CHIP_NO_ERROR on success
+     */
+    CHIP_ERROR Init(const chip::TLV::TLVReader & aReader);
```

Comment:
how come Init() isn't on ParserBase?

Reply:
Init is different from different message, 

Init(const chip::TLV::TLVReader & aReader)
for EventDataElement, in init, it  need to check if container is structure, chip::TLV::kTLVType_Structure
CHIP_ERROR EventDataElement::Parser::Init(const chip::TLV::TLVReader & aReader)
VerifyOrExit(chip::TLV::kTLVType_Structure == mReader.GetType(), err = CHIP_ERROR_WRONG_TLV_TYPE);

for commandPath, in init, it needs to check against kTLVType_Path
CHIP_ERROR CommandPath::Parser::Init(const chip::TLV::TLVReader & aReader)
VerifyOrExit(chip::TLV::kTLVType_Path == mReader.GetType(), err = CHIP_ERROR_WRONG_TLV_TYPE);